core/src/main/java/org/apache/calcite/rel/rules/SortRemoveRule.java

```diff
@@ -63,5 +83,27 @@ public SortRemoveRule(RelBuilderFactory relBuilderFactory) {
         .getTrait(RelCollationTraitDef.INSTANCE);
     final RelTraitSet traits = sort.getInput().getTraitSet().replace(collation);
     call.transformTo(convert(sort.getInput(), traits));
+    return true;
+  }
+
+  // Sort is removed if all of following conditions are met
+  // 1. Sort's input is guaranteed to produce at most one row
+  // 2. If fetch is greater than 0
+  // 3. If offset is less than 1
+  public static boolean shouldRemoveSortBasedOnRowCount(final RelMetadataQuery mq,
```

Comment:
it's hard to look at this method separately from the transformations of fetch / offset literal conversion - since that may change it to -1

I think it might worth to consider changing the signature to give this method the maxRowCount and the possible literals of fetch/offset

Reply:
@kgyrtkirk Make sense. I will update the code.